src/fairseq2/datasets/huggingface.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+from typing import Any, Callable, Dict, Optional
+
+from fairseq2.data.data_pipeline import (
+    Collater,
+    DataPipelineBuilder,
+    create_bucket_sizes,
+    read_sequence,
+)
+from fairseq2.datasets.data_reader import DataPipelineReader
+from fairseq2.gang import Gang
+from fairseq2.models.seq2seq import Seq2SeqBatch
+
+try:
+    from datasets import (  # type: ignore[attr-defined,import-untyped,import-not-found]
+        Dataset,
+        DatasetDict,
+    )
+except ImportError:
+    has_datasets = False
+else:
+    has_datasets = True
+
+
+Example = Dict[str, Any]
+
+
+class Batcher(abc.ABC):
```

Comment:
I suggest using the batching types defined in [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/datasets/batching.py) along with how they are used in other datasets (e.g. see [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/datasets/instruction.py#L332-L352)).